### PR TITLE
PATA support

### PR DESCRIPTION
**Pull request by brynolv**
This is mega-scuffed. I welcome ideas to get rid of the `print!`s.

Tested in both debug and release, R-W x2 and W-R x2.

### CODE REVIEW

File: src/harddisk/pata.rs [hunk 1]
```diff
@@ -1,8 +1,32 @@
-use crate::svec::SVec;
+use core::sync::atomic::AtomicBool;
 
+use x86_64::instructions::port::{Port, PortReadOnly, PortWriteOnly};
+
+use crate::{nop, svec::SVec};
```
Comment (Elekrisk) [inline src/harddisk/pata.rs]: `crate::nop` doesn't exist in `crate`

**Review by Elekrisk — CHANGES_REQUESTED** [review]
Only one thing pops out to(?) me.

File: src/harddisk/pata.rs (hunk 1 above)
Comment (brynolv) [reply src/harddisk/pata.rs]: Not *anymore*... and never in this repo. I'll run `rustfmt`as well, while I'm at it.

**Review by Elekrisk — APPROVED** [review]

**Review by Elekrisk — APPROVED** [review]

**Review by Elekrisk — APPROVED** [review]

**Review by Vikdos — APPROVED** [review]
ok:tm:

File: src/harddisk/pata.rs [hunk 2]
```diff
@@ -1,37 +1,229 @@
...
+        if status == 0 {
+            panic!("Drive does not exist");
+        }
+        while STATUS_REG.read() & 0x80 != 0 {}
+        if LBA_MID_REG.read() != 0 || LBA_HIGH_REG.read() != 0 {
+            panic!("Non-ATA drive");
+        }
+        loop {
+            let status = STATUS_REG.read();
+            if status & 8 == 0 {
```
Comment (Elekrisk) [inline src/harddisk/pata.rs]: Shouldn't this be `0x80`? according to oswiki

**Review by Elekrisk — CHANGES_REQUESTED** [review]
Noticed a thing :P

File: src/harddisk/pata.rs (hunk 2 above)
Comment (Elekrisk) [reply src/harddisk/pata.rs]: Actually, no, i read wrong :P

**Review by Elekrisk — APPROVED** [review]
nvm lol

File: src/harddisk/pata.rs (hunk 2 above)
Comment (Elekrisk) [reply src/harddisk/pata.rs]: I did read wrong, but there is still something wrong.
It should probs be `== 8`

**Review by Elekrisk — CHANGES_REQUESTED** [review]
There was something wrong lol

**Comment by brynolv** [comment]
Draft till LBA28 (what's in QEMU) is supported

**Comment by brynolv** [comment]
It works as well as it did before

**Review by Elekrisk — APPROVED** [review]

**Pull request merged**